src/Npgsql/NpgsqlTypes/NpgsqlRange.cs

```diff
@@ -131,49 +260,210 @@ internal NpgsqlRange([CanBeNull] T lowerBound, [CanBeNull] T upperBound, RangeFl
+
+            if (separator == -1)
+                throw new FormatException("Invalid format for PostgreSQL range type. Ranges must contain ','.");
+
+            if (separator != value.LastIndexOf(','))
+                // TODO: this should be replaced to handle quoted commas.
+                throw new NotSupportedException("Ranges with embedded commas are not currently supported.");
+
+            // Skip the opening bracket and stop short of the separator.
+            StringSegment lowerSegment = value.Subsegment(1, separator - 1);
+
+            // Skip past the separator and stop short of the closing bracket.
+            StringSegment upperSegment = value.Subsegment(separator + 1, value.Length - separator - 2);
+
+            var lowerInfinite =
+                !lowerSegment.HasValue ||
+                lowerSegment.Equals(string.Empty, StringComparison.OrdinalIgnoreCase) ||
+                lowerSegment.Equals(NullLiteral, StringComparison.OrdinalIgnoreCase) ||
+                lowerSegment.Equals(LowerInfinityLiteral, StringComparison.OrdinalIgnoreCase);
+
+            var upperInfinite =
+                !upperSegment.HasValue ||
+                upperSegment.Equals(string.Empty, StringComparison.OrdinalIgnoreCase) ||
+                upperSegment.Equals(NullLiteral, StringComparison.OrdinalIgnoreCase) ||
+                upperSegment.Equals(UpperInfinityLiteral, StringComparison.OrdinalIgnoreCase);
+
+            TypeConverter typeConverter = TypeDescriptor.GetConverter(typeof(T));
+
+            T lower = lowerInfinite ? default : (T)typeConverter.ConvertFromString(lowerSegment.Value);
+            T upper = upperInfinite ? default : (T)typeConverter.ConvertFromString(upperSegment.Value);
+
+            if (lowerInclusive && lowerInfinite)
+                // PostgreSQL automatically converts '[-infinity,tody]' to '(-infinity,tody]'.
+                lowerInclusive = false;
+
+            if (upperInclusive && upperInfinite)
+                // PostgreSQL automatically converts '[tody,infinity]' to '[today,infinity)'.
+                upperInclusive = false;
+
+            return new NpgsqlRange<T>(lower, lowerInclusive, lowerInfinite, upper, upperInclusive, upperInfinite);
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        /// Represents a type converter for <see cref="NpgsqlRange{T}" />.
+        /// </summary>
+        internal class RangeTypeConverter : TypeConverter
+        {
+            /// <inheritdoc />
+            public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceType)
+                => sourceType == typeof(string) || sourceType == typeof(StringSegment) || base.CanConvertFrom(context, sourceType);
+
+            /// <inheritdoc />
+            public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
+            {
+                switch (value)
+                {
+                case string s:
+                {
```

Comment:
These curlies are redundant, please remove.

Actually if you remove the `StringSegment` dependency the need for a switch may go away.